File: DOUBLE LINKED LISTS/Delete Operations/DeleteAtStart.java
public class DeleteAtStart {

    // ======================================================
    // NODE CLASS + HELPERS
    // ======================================================

    static class Node {
        int data;
        Node prev;
        Node next;
    }

    static Node createNode(int data) {
        Node newNode = new Node();
        newNode.data = data;
        newNode.prev = null;
        newNode.next = null;
        return newNode;
    }

    static void traverse(Node head) {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }


    // ======================================================
    // DELETE AT THE BEGINNING
    // ======================================================

    static Node deleteAtBeginning(Node head) {

        /*
            ─────────────────────────────────────────
            CASE 1: The list is empty.
            ─────────────────────────────────────────
        */

        if (head == null) {
            System.out.println("List is empty. Nothing to delete.");
            return head;
        }


        /*
            ─────────────────────────────────────────
            CASE 2: The list has one or more nodes.

                    head
                     ↓
            null ← [10] ⇄ [20] ⇄ [30] → null

            We want to remove [10] and make [20] the new head.


            In C this needed FOUR steps:

                struct Node *temp = head;   // 1. save address for free()
                head = head->next;          // 2. move head forward
                if (head != NULL)           // 3. clear new head's prev
                    head->prev = NULL;      //    (guarded — see below)
                free(temp);                 // 4. release memory


            In Java, steps 1 and 4 vanish (no free(), no temp).
            Step 2 is one line:

                head = head.next;

            Step 3 STILL EXISTS and still needs its guard:

                if (head != null)
                    head.prev = null;

            ⚠ THE DLL-SPECIFIC STEP: [20].prev still points at the
            deleted [10]. If we leave it that way, rule 1 breaks and
            anyone walking backwards from head lands on garbage.
            The guard matters when the list had exactly ONE node:
            after step 2, head is null, and writing head.prev would
            throw a NullPointerException.

            The deleted node becomes unreachable → GC reclaims it.
        */

        head = head.next;              // STEP 2: move head forward

        if (head != null) {            // STEP 3: clear prev — guarded!
            head.prev = null;
        }

        return head;
    }


    // ======================================================
    // MAIN
    // ======================================================

    public static void main(String[] args) {

        /*
            Build:  null ← [10] ⇄ [20] ⇄ [30] → null
        */

        Node n1 = createNode(10);
        Node n2 = createNode(20);
        Node n3 = createNode(30);

        n1.next = n2;
        n2.prev = n1;
        n2.next = n3;
        n3.prev = n2;

        Node head = n1;

        System.out.print("Original list:           ");
        traverse(head);              // 10 20 30

        /*
            Delete the first node.

            Before:   null ← [10] ⇄ [20] ⇄ [30]
            After:    null ← [20] ⇄ [30]
        */

        head = deleteAtBeginning(head);

        System.out.print("After deleting at start: ");
        traverse(head);              // 20 30

        /*
            Delete again.
        */

        head = deleteAtBeginning(head);

        System.out.print("After deleting at start: ");
        traverse(head);              // 30

        /*
            Delete the last remaining node.
            Exercises the `if (head != null)` guard — after this
            deletion head is null and there is no prev to clear.
        */

        head = deleteAtBeginning(head);

        System.out.print("After deleting at start: ");
        traverse(head);              // (blank line — empty list)

        /*
            Try to delete from an empty list.
            Should print a message and do nothing.
        */

        head = deleteAtBeginning(head);
    }
}

File: DOUBLE LINKED LISTS/Delete Operations/DeleteAtStart.c
#include <stdio.h>
#include <stdlib.h>


// ==============================
// NODE STRUCTURE + HELPERS
// ==============================

struct Node {
    int data;
    struct Node *prev;
    struct Node *next;
};

struct Node *createNode(int data) {
    struct Node *newNode = malloc(sizeof(struct Node));
    if (newNode == NULL) return NULL;
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

void traverse(struct Node *head) {
    struct Node *current = head;
    while (current != NULL) {
        printf("%d ", current->data);
        current = current->next;
    }
    printf("\n");
}


// ==============================
// DELETE AT THE BEGINNING
// ==============================

struct Node *deleteAtBeginning(struct Node *head) {

    /*
        ─────────────────────────────────────────
        CASE 1: The list is empty.
        ─────────────────────────────────────────

            head → NULL

        Nothing to delete. Return head unchanged.
    */

    if (head == NULL) {
        printf("List is empty. Nothing to delete.\n");
        return head;
    }


    /*
        ─────────────────────────────────────────
        CASE 2: The list has one or more nodes.
        ─────────────────────────────────────────

                head
                 ↓
        NULL ← [10] ⇄ [20] ⇄ [30] → NULL

        We want to remove [10] and make [20] the new head.

        STEP 1: Save the old head so we can free it later.

            struct Node *temp = head;

                temp,head
                    ↓
        NULL ← [10] ⇄ [20] ⇄ [30] → NULL


        STEP 2: Move head forward.

            head = head->next;

                 temp     head
                   ↓        ↓
        NULL ← [10] ⇄ [20] ⇄ [30] → NULL

        ⚠ NEW DLL STEP — the one the SLL version didn't have:
        [20].prev still points at the deleted [10].
        If we leave it that way, rule 1 breaks and anyone who walks
        backwards from head would land on freed memory.


        STEP 3: Clear the new head's prev — but ONLY if a new
        head exists!

            if (head != NULL)
                head->prev = NULL;

            Why the guard? If the list had exactly ONE node,
            then after step 2, head is NULL — and writing
            head->prev would crash. This guard is the DLL's
            extra edge case compared to delete-at-start in an SLL.


        STEP 4: Free the old node (C only).

            free(temp);

        State:

                  head
                   ↓
        NULL ← [20] ⇄ [30] → NULL

        Boundary check: head->prev == NULL ✓ (rule 1 restored).
    */

    struct Node *temp = head;      // STEP 1: save the address of [10]
    head = head->next;             // STEP 2: move head forward to [20]

    if (head != NULL) {            // STEP 3: clear prev — guarded!
        head->prev = NULL;
    }

    free(temp);                    // STEP 4: release the old [10]

    return head;
}


// ==============================
// MAIN
// ==============================

int main() {

    /*
        Build:  NULL ← [10] ⇄ [20] ⇄ [30] → NULL
    */

    struct Node *n1 = createNode(10);
    struct Node *n2 = createNode(20);
    struct Node *n3 = createNode(30);

    n1->next = n2;
    n2->prev = n1;
    n2->next = n3;
    n3->prev = n2;

    struct Node *head = n1;

    printf("Original list:           ");
    traverse(head);              // 10 20 30

    /*
        Delete the first node.

        Before:   NULL ← [10] ⇄ [20] ⇄ [30]
        After:    NULL ← [20] ⇄ [30]
    */

    head = deleteAtBeginning(head);

    printf("After deleting at start: ");
    traverse(head);              // 20 30

    /*
        Delete again.
    */

    head = deleteAtBeginning(head);

    printf("After deleting at start: ");
    traverse(head);              // 30

    /*
        Delete the last remaining node.
        Exercises the `if (head != NULL)` guard — after this
        deletion head is NULL and there is no prev to clear.
    */

    head = deleteAtBeginning(head);

    printf("After deleting at start: ");
    traverse(head);              // (blank line — empty list)

    /*
        Try to delete from an empty list.
        Should print a message and do nothing.
    */

    head = deleteAtBeginning(head);

    return 0;
}
